### PR TITLE
Convert tool.js to typescript (based off #4166)

### PR DESCRIPTION
**Pull request by mattgodbolt**
Based off #4166 with the hard original work by @MarkusJx . This is work in progress on fixing the underlying code and then trying to pull that into Markus's changes (with some modifications to not paper over missing types).

### CODE REVIEW

**Comment by jeremy-rifkin** [comment]
It looks like the "The current language or compiler does not support this tool" banner is broken, haven't spotted anything else yet but it's hard to test locally

**Comment by mattgodbolt** [comment]
Right - this is a tricky one :( Any thoughts welcomed on how best to continue.

File: static/panes/compiler.ts [hunk 1]
```diff
@@ -415,7 +415,13 @@ export class Compiler extends MonacoPane<monaco.editor.IStandaloneCodeEditor, Co
                     readOnly: true,
                     language: 'asm',
                     glyphMargin: !options.embedded,
-                    // guides: false,
+                    guides: {
```
Comment (jeremy-rifkin) [inline static/panes/compiler.ts:418]: changed this while I was here, I think this is what we want

File: static/event-map.ts [hunk 2]
```diff
@@ -75,7 +75,7 @@ export type EventMap = {
     editorDisplayFlow: (editorId: number, flow: MessageWithLocation[]) => void;
     editorLinkLine: (editorId: number, lineNumber: number, colBegin: number, colEnd: number, reveal: boolean) => void;
     editorOpen: (editorId: number) => void;
-    editorSetDecoration: (editorId: number, lineNumber: number, reveal: boolean) => void;
+    editorSetDecoration: (editorId: number, lineNumber: number, reveal: boolean, column?: number) => void;
```
Comment (jeremy-rifkin) [inline static/event-map.ts:78]: Updated the event type here

File: static/panes/tool.ts [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+        this.compilerInfo.treeId = treeId;
+        this.updateTitle();
+    }
+
+    onCompileResult(id, compiler, result) {
+        try {
+            if (id !== this.compilerInfo.compilerId) return;
+            if (compiler) this.compilerInfo.compilerName = compiler.name;
+
+            const foundTool = _.find(compiler.tools, tool => tool.tool.id === this.toolId);
```
Comment (jeremy-rifkin) [inline static/panes/tool.ts:418]: Wasn't confident enough to try to remove the `_.find` uses here, wanted to minimize potential problems introduced by this pr

File: static/components.interfaces.ts [hunk 4]
```diff
@@ -110,17 +110,17 @@ export type OutputState = StateWithTree & {
...
-export type ToolViewState = StateWithTree & {
-    compiler: number; // CompilerId
-    editor: number; // EditorId
-    toolId: string;
-    args: unknown;
-    monacoStdin: boolean;
-};
+export type ToolViewState = StateWithTree &
+    ToolState & {
+        id: number; // CompilerID (TODO: Why is this not part of StateWithTree)
```
Comment (mattgodbolt) [inline static/components.interfaces.ts:115]: if the TODO is actionable can we tie it to an issue? or if it's just a "here's a weird thing" let's drop the TODO part?

**Review by mattgodbolt — COMMENTED** [review]
Looks good to me... I think unless you have any issues we should "just" merge and handle the fallout. (I can't approve as I made the PR!)
Thanks so much!

File: static/components.interfaces.ts (hunk 4 above)
Comment (jeremy-rifkin) [reply static/components.interfaces.ts:115]: Good idea, will make an issue for this

**Comment by jeremy-rifkin** [comment]
Agreed, let's get this merged and fix any bugs that pop up later :)

**Comment by jeremy-rifkin** [comment]
🎉

**Pull request merged**